### DISCUSSION TITLE
[tables] Draggable table rows

### DISCUSSION THREAD

**Discussion by ArvidDeJong** [livewire/flux#209] [Feature Requests]
### Sorting Table Rows with SortableJS

Currently, I am using [SortableJS](https://github.com/SortableJS/Sortable) to handle sorting of table rows in my project. SortableJS is a powerful and flexible JavaScript library that integrates easily, but it would be great if this functionality were built into [Flux](https://fluxui.dev) by default.

Having this feature natively available in Flux would streamline the implementation process and reduce the need for external libraries.

**Comment by ArvidDeJong** [comment]
Just an example of how i sort now

https://github.com/user-attachments/assets/eb4e4c99-a9d8-4469-b8f8-30f90d1efd8d

**Comment by colinmac17** [comment]
FWIW, this is what I'm doing now with Sortable JS and it works really well:

```js
 @script
    <script>
        new window.sortable.create(document.getElementById('video-sortable'), {
            handle: '.handle', // handle's class
            animation: 150,
            onUpdate: function (event) {
                // Get all rows in the table
                const rows = event.target.getElementsByTagName('tr');

                // Convert to array and map through to get datasets
                const items = Array.from(rows).map(row => {
                    // If you need specific data attributes:
                    return {
                        id: row.dataset.videoId,
                    };
                });
                
                // If you need the new order with specific attributes:
                const newOrder = items.map(item => item.id); // Gets array of just IDs

                // Call wire method to update order
                $wire.updateVideoOrder(newOrder);
            },
        });
    </script>
    @endscript
```

```php
//livewire
 public function updateVideoOrder(array $videoIds): void
    {
        foreach($videoIds as $index => $videoId){
            ProfileVideo::query()
                ->where('id', $videoId)
                ->update(['order_rank' => $index + 1]);
        }
    }
```

**Reply by igmltd** [reply]
Just to add, Alpine has a sort wrapper with a clean api

https://alpinejs.dev/plugins/sort

**Reply by colinmac17** [reply]
Ah thank you! Even better

**Reply by ArvidDeJong** [reply]
Thanks, i will use this one for now, but it would still be nice if the comes a shipped option :draggable in tables :)

```<flux:rows :draggable>```

**Comment by Hesesses** [comment]
Would be awesome to have sortable supported for tables and nav items 🚀

**Comment by craigcordell** [comment]
﻿﻿
#  Drag-and-Drop Sorting with Flux + Alpine Sort

Flux support for drag-and-drop sorting would be a great enhancement. For now, here’s how I implemented it using the [Alpine Sort plugin](https://alpinejs.dev/plugins/sort).

> **Note:** The Alpine Sort plugin only provides the moved item's ID and new position — it does not return the full list of reordered IDs.  
> In this setup, new items default to `sort = 0`, and reordering starts from `1`, leaving space at the top for new entries.

---

##  Install the Sort Plugin

```bash
npm install alpinejs @alpinejs/sort
```

---

##  Register the Plugin in `resources/js/app.js`

```js
import sort from '@alpinejs/sort';

document.addEventListener('alpine:init', () => {
    Alpine.plugin(sort);
});
```

---

##  Blade Example (Flux Table)

Note: `x-sort` goes on `<flux:table.rows>`, **not** the outer `<flux:table>`.

```blade
<flux:table>
    <flux:table.columns>
        <flux:table.column>To do</flux:table.column>
    </flux:table.columns>

    <flux:table.rows x-sort="$wire.updateOrder($item, $position)">
        @foreach ($todos as $todo)
            <flux:table.row x-sort:item="{{ $todo->id }}" class="cursor-ns-resize">
                <flux:table.cell>{{ $todo->task }}</flux:table.cell>
            </flux:table.row>
        @endforeach
    </flux:table.rows>
</flux:table>
```

---

##  Livewire Component Method

```php
public function updateOrder($id, $newPosition)
{
    $todos = $this->todos;

    $moving = $todos->firstWhere('id', $id);
    $todos = $todos->reject(fn($todo) => $todo->id == $id);

    $todos->splice($newPosition, 0, [$moving]);

    $todos->values()->each(function ($todo, $index) {
        $todo->update(['sort' => $index + 1]);
    });

    $this->mount(); // Reload $this->todos sorted by the `sort` column
}
```
